### ISSUE TITLE
Rook creates OSD on node with no spare block devices

### ISSUE THREAD

**Issue by jpds** [rook/rook#1007]
Hello,

I've deployed rook with:

```
  storage:                # cluster level storage configuration and selection
    useAllNodes: true
    useAllDevices: true
```

This worked fine with my first three nodes when it took the `sdb`s from those nodes and created a cluster fine.

I then added a node with two disks that were in an `mdadm` RAID with LVM on that for the system partitions, but no spare block devices. I expected nothing to be added from Rook to this system. However to my surprise, it had decided to create one on the filesystem on the LVM:

```
Pod:  rook-ceph-osd-l7xxq
Node: node4
-ne -osd3 
/dev/mapper/system-root 912.1G filestore
```

This is also a filestore instead of the bluestore I specified.

Here are the initial logs from the container:

```
[WARN  tini (7)] Tini is not running as PID 1 and isn't registered as a child subreaper.
Zombie processes will not be re-parented to Tini, so zombie reaping won't work.
To fix the problem, use the -s option or set the environment variable TINI_SUBREAPER to register Tini as a child subreaper, or run Tini as PID 1.
2017-09-23 07:29:52.475366 I | rook: starting Rook v0.5.0-105.gc3b000f with arguments '/usr/local/bin/rook osd'
2017-09-23 07:29:52.475449 I | rook: flag values: --admin-secret=*****, --ceph-config-override=/etc/rook/config/override.conf, --cluster-name=rook, --config-dir=/var/lib/rook, --data-device-filter=all, --data-devices=, --data-directories=, --force-format=false, --fsid=, --help=false, --location=, --log-level=INFO, --metadata-device=, --mon-endpoints=rook-ceph-mon0=...,rook-ceph-mon1=...,rook-ceph-mon2=..., --mon-secret=*****, --node-name=node4, --osd-database-size=20480, --osd-journal-size=5120, --osd-store=bluestore, --osd-wal-size=576, --private-ipv4=**.**.**.**, --public-ipv4=**.**.**.**
2017-09-23 07:29:52.475456 I | cephmon: parsing mon endpoints: rook-ceph-mon0=...,rook-ceph-mon1=...,rook-ceph-mon2=...
2017-09-23 07:29:52.475472 I | proc: Running process hostname with args: [node4]
2017-09-23 07:29:52.475478 I | exec: Running command: hostname node4
2017-09-23 07:29:52.476658 I | cephmon: writing config file /var/lib/rook/rook/rook.config
2017-09-23 07:29:52.476787 I | cephmon: generated admin config in /var/lib/rook/rook
2017-09-23 07:29:52.476792 I | cephosd: discovering hardware
2017-09-23 07:29:52.476840 I | exec: Running command: lsblk --all --noheadings --list --output KNAME
2017-09-23 07:29:52.478614 I | exec: Running command: lsblk /dev/sda --bytes --nodeps --pairs --output SIZE,ROTA,RO,TYPE,PKNAME
2017-09-23 07:29:52.479618 I | exec: Running command: sgdisk --print /dev/sda
2017-09-23 07:29:52.481122 I | exec: Running command: df --output=source,fstype
2017-09-23 07:29:52.481784 I | exec: Running command: lsblk /dev/sda1 --bytes --nodeps --pairs --output SIZE,ROTA,RO,TYPE,PKNAME
2017-09-23 07:29:52.482685 I | exec: Running command: df --output=source,fstype
2017-09-23 07:29:52.483344 I | exec: Running command: lsblk /dev/sdb --bytes --nodeps --pairs --output SIZE,ROTA,RO,TYPE,PKNAME
2017-09-23 07:29:52.484289 I | exec: Running command: sgdisk --print /dev/sdb
2017-09-23 07:29:52.485674 I | exec: Running command: df --output=source,fstype
2017-09-23 07:29:52.486665 I | exec: Running command: lsblk /dev/sdb1 --bytes --nodeps --pairs --output SIZE,ROTA,RO,TYPE,PKNAME
2017-09-23 07:29:52.487696 I | exec: Running command: df --output=source,fstype
2017-09-23 07:29:52.488386 I | exec: Running command: lsblk /dev/md0 --bytes --nodeps --pairs --output SIZE,ROTA,RO,TYPE,PKNAME
2017-09-23 07:29:52.489466 I | exec: Running command: lsblk /dev/md0 --bytes --nodeps --pairs --output SIZE,ROTA,RO,TYPE,PKNAME
2017-09-23 07:29:52.490552 I | exec: Running command: lsblk /dev/dm-0 --bytes --nodeps --pairs --output SIZE,ROTA,RO,TYPE,PKNAME
2017-09-23 07:29:52.498242 I | exec: Running command: lsblk /dev/dm-0 --bytes --nodeps --pairs --output SIZE,ROTA,RO,TYPE,PKNAME
2017-09-23 07:29:52.499270 I | exec: Running command: lsblk /dev/dm-1 --bytes --nodeps --pairs --output SIZE,ROTA,RO,TYPE,PKNAME
2017-09-23 07:29:52.500149 I | exec: Running command: lsblk /dev/dm-1 --bytes --nodeps --pairs --output SIZE,ROTA,RO,TYPE,PKNAME
2017-09-23 07:29:52.500989 I | cephosd: creating and starting the osds
2017-09-23 07:29:52.501003 I | exec: Running command: lsblk /dev/sda --bytes --pairs --output NAME,SIZE,TYPE,PKNAME
2017-09-23 07:29:52.501976 I | exec: Running command: blkid /dev/sda1 -s PARTLABEL -o value
2017-09-23 07:29:52.502863 I | exec: Running command: df --output=source,fstype
2017-09-23 07:29:52.503524 I | cephosd: skipping device sda that is in use (not by rook). fs: , ownPartitions: false
2017-09-23 07:29:52.503539 I | exec: Running command: lsblk /dev/sdb --bytes --pairs --output NAME,SIZE,TYPE,PKNAME
2017-09-23 07:29:52.504511 I | exec: Running command: blkid /dev/sdb1 -s PARTLABEL -o value
2017-09-23 07:29:52.540523 I | exec: Running command: df --output=source,fstype
2017-09-23 07:29:52.543909 I | cephosd: skipping device sdb that is in use (not by rook). fs: , ownPartitions: false
2017-09-23 07:29:52.543987 I | cephosd: configuring osd devices: &{Entries:map[]}
2017-09-23 07:29:52.544086 I | cephosd: configuring osd dirs: map[/var/lib/rook:-1]
2017-09-23 07:29:52.544567 I | exec: Running command: ceph osd create 155ce113-167e-43ac-8d25-cbb796811586 --cluster=rook --conf=/var/lib/rook/rook/rook.config --keyring=/var/lib/rook/rook/client.admin.keyring --format json --out-file /tmp/703787263
2017-09-23 07:30:06.832755 I | cephosd: successfully created OSD 155ce113-167e-43ac-8d25-cbb796811586 with ID 3
2017-09-23 07:30:06.832986 I | cephmon: writing config file /var/lib/rook/osd3/rook.config
2017-09-23 07:30:06.833135 I | exec: Running command: ceph mon getmap --cluster=rook --conf=/var/lib/rook/rook/rook.config --keyring=/var/lib/rook/rook/client.admin.keyring --format json --out-file /tmp/206530642
2017-09-23 07:30:07.265496 I | exec: got monmap epoch 3
2017-09-23 07:30:07.265624 I | cephosd: Initializing OSD 3 file system at /var/lib/rook/osd3...
2017-09-23 07:30:07.265735 I | proc: Running process ceph-osd with args: [--mkfs --mkkey --id=3 --cluster=rook --conf=/var/lib/rook/osd3/rook.config --osd-data=/var/lib/rook/osd3 --osd-uuid=155ce113-167e-43ac-8d25-cbb796811586 --monmap=/var/lib/rook/osd3/tmp/activate.monmap --osd-journal=/var/lib/rook/osd3/journal --keyring=/var/lib/rook/osd3/keyring]
2017-09-23 07:30:07.265749 I | exec: Running command: ceph-osd --mkfs --mkkey --id=3 --cluster=rook --conf=/var/lib/rook/osd3/rook.config --osd-data=/var/lib/rook/osd3 --osd-uuid=155ce113-167e-43ac-8d25-cbb796811586 --monmap=/var/lib/rook/osd3/tmp/activate.monmap --osd-journal=/var/lib/rook/osd3/journal --keyring=/var/lib/rook/osd3/keyring
2017-09-23 07:30:07.285315 I | mkfs-osd3: 2017-09-23 07:30:07.285245 7f74e4e92e00  0 ceph version ec053b628 (cec053b62872f1342c8ebd8bc6d5e9c21aa82a1e) luminous (rc), process (unknown), pid 86
2017-09-23 07:30:07.451903 I | mkfs-osd3: 2017-09-23 07:30:07.451714 7f74e4e92e00  0 filestore(/var/lib/rook/osd3) backend generic (magic 0xef53)
2017-09-23 07:30:07.720135 I | mkfs-osd3: 2017-09-23 07:30:07.719943 7f74e4e92e00 -1 journal FileJournal::_open: disabling aio for non-block journal.  Use journal_force_aio to force use of aio anyway
2017-09-23 07:30:07.891906 I | mkfs-osd3: 2017-09-23 07:30:07.891742 7f74e4e92e00  0 filestore(/var/lib/rook/osd3) mkjournal(1070): created journal on /var/lib/rook/osd3/journal
2017-09-23 07:30:07.950935 I | mkfs-osd3: 2017-09-23 07:30:07.950781 7f74e4e92e00  0 filestore(/var/lib/rook/osd3) backend generic (magic 0xef53)
2017-09-23 07:30:07.951444 I | mkfs-osd3: 2017-09-23 07:30:07.951360 7f74e4e92e00  0 genericfilestorebackend(/var/lib/rook/osd3) detect_features: FIEMAP ioctl is disabled via 'filestore fiemap' config option
2017-09-23 07:30:07.951468 I | mkfs-osd3: 2017-09-23 07:30:07.951380 7f74e4e92e00  0 genericfilestorebackend(/var/lib/rook/osd3) detect_features: SEEK_DATA/SEEK_HOLE is disabled via 'filestore seek data hole' config option
2017-09-23 07:30:07.951477 I | mkfs-osd3: 2017-09-23 07:30:07.951383 7f74e4e92e00  0 genericfilestorebackend(/var/lib/rook/osd3) detect_features: splice() is disabled via 'filestore splice' config option
2017-09-23 07:30:11.035478 I | mkfs-osd3: 2017-09-23 07:30:11.035358 7f74e4e92e00  0 genericfilestorebackend(/var/lib/rook/osd3) detect_features: syncfs(2) syscall fully supported (by glibc and kernel)
2017-09-23 07:30:11.036002 I | mkfs-osd3: 2017-09-23 07:30:11.035933 7f74e4e92e00  0 filestore(/var/lib/rook/osd3) limited size xattrs
2017-09-23 07:30:11.036451 I | mkfs-osd3: 2017-09-23 07:30:11.036389 7f74e4e92e00  0 filestore(/var/lib/rook/osd3) start omap initiation
...
```

**Comment by jpds** [comment]
I've just experimented with:

```
  storage:                # cluster level storage configuration and selection
    useAllNodes: true
    useAllDevices: true
    deviceFilter: '^sd.'
```

And:

```
  storage:                # cluster level storage configuration and selection
    useAllNodes: true
    useAllDevices: false
    deviceFilter: '^sd.'
```

But for some reason, rook insists on creating an OSD on the node with no drives in it spare.

**Comment by jbw976** [comment]
Hi @jpds, thanks for reporting this issue.  After looking at the code, I think I understand why it is behaving this way.  On the following lines, it looks like if Rook was unable to find/use any available devices, and no directories were specified either, then it'll use default to creating an OSD in the default `/var/lib/rook` directory:
https://github.com/rook/rook/blob/a4a1a98e2bbe795bb8cdde2d9546a1ebbb92535d/pkg/ceph/osd/daemon.go#L73
https://github.com/rook/rook/blob/a4a1a98e2bbe795bb8cdde2d9546a1ebbb92535d/pkg/ceph/osd/daemon.go#L184

Perhaps this logic should be updated so that if the user tried to specify devices, but no available ones were found, then no OSD should be created at all.  The users intent to use devices could not be satisfied, so do nothing.

Does that sound like the behavior you are looking for?

**Comment by jpds** [comment]
Hello @jbw976 - that sounds very close. There is also the case where I specify `useAllNodes` and `useAllDevices` to `true` and there are no block devices without a local filesystem on the node - in this scenerio Rook shouldn't provision anything on the node either.

**Comment by jbw976** [comment]
that's correct @jpds, thanks. I'm actually touching this logic right now while working on #726 and it's a very simple change, so I'll go ahead and make it now.  There's a lot of testing that needs to be done for #726 so the testing on this small change will be amortized in that effort.  Thanks!

**Comment by jbw976** [comment]
@jpds a fix for this has been merged to master with #1024.  let me know if you still see this behavior on any build after that merge. Thanks again for reporting this!

**Comment by EdSchouten** [comment]
It looks like this issue hasn't been resolved at all. We still see Rook spinning up filestores on systems without any spare SSDs. This is with Rook 0.9.
